### PR TITLE
Search

### PR DESCRIPTION
**Pull request by YuriHechter**

### CODE REVIEW

File: es/content_units_index.go [hunk 1]
```diff
@@ -326,6 +325,9 @@ func (index *ContentUnitsIndex) prepareIndexUnit(cu *mdbmodels.ContentUnit, inde
 			if val, ok := indexData.Sources[cu.UID]; ok {
 				unit.FilterValues = append(unit.FilterValues, KeyValues(consts.ES_UID_TYPE_SOURCE, val)...)
 				unit.TypedUids = append(unit.TypedUids, KeyValues(consts.ES_UID_TYPE_SOURCE, val)...)
+				//  We dont add TitleSuggest to CU with source
+			} else {
+				unit.TitleSuggest = Suffixes(i18n.Name.String)
```
Comment (bbfsdev) [inline es/content_units_index.go:330]: Are we skipping title suggest for a unit if it is not a source?

File: es/sources_index.go [hunk 2]
```diff
@@ -360,9 +361,31 @@ func (index *SourcesIndex) indexSource(mdbSource *mdbmodels.Source, parents []st
 			}
 			authors := authorsByLanguage[i18n.Language]
 			s := append(authors, pathNames...)
-			source.Title = s[len(s)-1]
+			leaf := s[len(s)-1]
+			source.Title = leaf
```
Comment (bbfsdev) [inline es/sources_index.go:365]: Using only the name of the article? Without parents?

File: es/sources_index.go [hunk 3]
```diff
@@ -360,9 +361,31 @@ func (index *SourcesIndex) indexSource(mdbSource *mdbmodels.Source, parents []st
...
+			suffixes := Suffixes(strings.Join(s, " "))
+			if len(s) > 2 {
+				suffixes = append(suffixes, ConcateFirstToLast(s))
```
Comment (bbfsdev) [inline es/sources_index.go:368]: Suffixes here will also include:
<Author> <Article Name>
?

File: consts/consts.go [hunk 4]
```diff
@@ -590,3 +591,11 @@ var LATENCY_LOG_OPERATIONS_FOR_SEARCH = []string{
 	LAT_DOSEARCH_MULTISEARCHTWEETSDO,
 	LAT_DOSEARCH_TYPOSUGGESTDO,
 }
+
+const (
+	SRC_SHAMATI = 141
+)
+
+var ES_SRC_PARENTS_FOR_CHAPTER_POSITION_INDEX = map[int64]bool{
+	SRC_SHAMATI: true,
```
Comment (bbfsdev) [inline consts/consts.go:600]: Only for Shamati?
Nothing else?

File: utils/languages.go [hunk 5]
```diff
@@ -177,3 +179,49 @@ func DetectLanguage(text string, interfaceLanguage string, acceptLanguage string
 	log.Debug("DetectLanguage: using default language")
 	return consts.SEARCH_LANG_ORDER[interfaceLanguage]
 }
+
+func NumberInHebrew(n int) string {
```
Comment (bbfsdev) [inline utils/languages.go:183]: Nice. Please add test for this.

Comment (bbfsdev) [reply utils/languages.go:183]: What is people write:
ט"ז

I mean with quotes... will it work?

File: search/variables/convention_location.variable [hunk 6]
```diff
@@ -168,86 +168,86 @@ he,United States|New Jersey => ניו גרזי
...
-# ru,Ukraine|Kiev => 
-# ru,Ukraine|Odessa => 
-# ru,United Kingdom|Blackpool => 
-# ru,United States|Buffalo => 
-# ru,United States|Los Angeles => 
-# ru,United States|Miami => 
-# ru,United States|New Jersey => 
-# ru,United States|New York => 
-# ru,United States|San Francisco => 
+ru,Argentina => Аргентина
```
Comment (bbfsdev) [inline search/variables/convention_location.variable:171]: Nice.

File: es/synonyms/en.txt [hunk 7]
```diff
@@ -0,0 +1,109 @@
+# Authors
```
Comment (bbfsdev) [inline es/synonyms/en.txt:1]: Nice

Comment (bbfsdev) [reply es/synonyms/en.txt:1]: How did you check this works well?

File: es/synonyms/ru.txt [hunk 8]
```diff
@@ -0,0 +1,156 @@
+# authors
```
Comment (bbfsdev) [inline es/synonyms/ru.txt:1]: Nice, How did you check this works well?

File: misc/demo.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 import subprocess
 import threading
 import urlparse
+import sys
```
Comment (bbfsdev) [inline misc/demo.py:16]: Those are only my changes, right?

File: search/engine.go [hunk 10]
```diff
@@ -178,10 +178,16 @@ func (e *ESEngine) GetSuggestions(ctx context.Context, query Query, preference s
 	// Run grammar suggestions in parallel.
 	grammarSuggestionsChannel := make(chan map[string][]VariablesByPhrase)
 	go func() {
+		defer func() {
```
Comment (bbfsdev) [inline search/engine.go:181]: Not sure what is that for?

File: search/engine.go [hunk 11]
```diff
@@ -483,6 +515,12 @@ func (e *ESEngine) DoSearch(ctx context.Context, query Query, sortBy string, fro
 	// Seach intents and grammars in parallel to native search.
 	intentsChannel := make(chan []Intent)
 	go func() {
+		defer func() {
```
Comment (bbfsdev) [inline search/engine.go:518]: You use this pattern of inserting empty result for recover...
What recover does?

What that comes to fix? some kind of deadlock when errors are happening?

File: search/grammars/landing-pages.grammar [hunk 12]
```diff
@@ -228,7 +228,36 @@ ru,articles => статьи
 ru,blog => блог
 ru,clips => клипы
 ru,clips => Клип
+ru,conventions => конгресс
```
Comment (bbfsdev) [inline search/grammars/landing-pages.grammar:231]: Nice!

File: es/content_units_index.go (hunk 1 above)
Comment (YuriHechter) [reply es/content_units_index.go:330]: yes, if the CU is related to a source then we skip the suggest. The suggest for the source itself is enough to find the CU also.

File: es/sources_index.go (hunk 2 above)
Comment (YuriHechter) [reply es/sources_index.go:365]: the parents are in the full_title field

File: es/sources_index.go (hunk 3 above)
Comment (YuriHechter) [reply es/sources_index.go:368]: it resolves problems with search like הרב"ש+[Article name]. 'Rabash' is the first node and article name is the last.

File: consts/consts.go (hunk 4 above)
Comment (YuriHechter) [reply consts/consts.go:600]: Search for Shamati was the major problem. Later we will consider adding more sources.

File: utils/languages.go (hunk 5 above)
Comment (YuriHechter) [reply utils/languages.go:183]: ט"ז = > works.
I will give Evgeny the task for creating tests. Anyway currently this function works100% ok.

File: es/synonyms/ru.txt (hunk 8 above)
Comment (YuriHechter) [reply es/synonyms/ru.txt:1]: we checked with Menuchin

File: misc/demo.py (hunk 9 above)
Comment (YuriHechter) [reply misc/demo.py:16]: yes

File: search/engine.go (hunk 11 above)
Comment (YuriHechter) [reply search/engine.go:518]: Go-Routens protection.
Without this pattern, panics inside go-routens are not handled and the whole program crashes.

File: search/engine.go (hunk 10 above)
Comment (YuriHechter) [reply search/engine.go:181]: see explanation below

**Pull request merged**